packages/service/src/service/base.ts

```diff
@@ -0,0 +1,115 @@
+import { isDomain } from '../utils';
+import { BtcAssetsApiError, ErrorCodes } from '../error';
+import { BaseApis, BaseApiRequestOptions, BtcAssetsApiToken } from '../types';
+
+export class BtcAssetsApiBase implements BaseApis {
+  public url: string;
+  public app?: string;
+  public domain?: string;
+  public origin?: string;
+  private token?: string;
+
+  constructor(props: { url: string; app?: string; domain?: string; origin?: string; token?: string }) {
+    this.url = props.url;
+    this.app = props.app;
+    this.domain = props.domain;
+    this.origin = props.origin;
+    this.token = props.token;
+
+    // Validation
+    if (this.domain && !isDomain(this.domain, true)) {
+      throw BtcAssetsApiError.withComment(ErrorCodes.ASSETS_API_INVALID_PARAM, 'domain');
+    }
+  }
+
+  async request<T>(route: string, options?: BaseApiRequestOptions): Promise<T> {
+    const { requireToken = true, method = 'GET', headers, params, ...otherOptions } = options ?? {};
+    if (requireToken && !this.token && !(this.app && this.domain)) {
+      throw BtcAssetsApiError.withComment(ErrorCodes.ASSETS_API_INVALID_PARAM, 'app, domain');
+    }
+    if (requireToken && !this.token) {
+      await this.init();
+    }
+
+    const packedParams = params ? '?' + new URLSearchParams(params).toString() : '';
+    const withOriginHeaders = this.origin ? { origin: this.origin } : void 0;
+    const withAuthHeaders = requireToken && this.token ? { Authorization: `Bearer ${this.token}` } : void 0;
+    const res = await fetch(`${this.url}${route}${packedParams}`, {
+      method,
+      headers: {
+        ...withOriginHeaders,
+        ...withAuthHeaders,
+        ...headers,
+      },
+      ...otherOptions,
+    } as RequestInit);
+
+    const status = res.status;
+
+    let text: string | undefined;
+    let json: Record<string, any> | undefined;
+    let ok: boolean = false;
+    try {
+      text = await res.text();
+      json = JSON.parse(text);
+      ok = json?.ok ?? res.ok ?? false;
+    } catch {
+      // do nothing
+    }
+
+    if (!json) {
+      if (status === 200) {
```

Comment:
Recommended to judge the status code after the response is returned, rather than after JSON parse

Reply:
Moved it before the JSON parse, and blocks it if status !== 200, do you think it's better?